### PR TITLE
Migrate the public endpoint Get DAG Stats to FastAPI

### LINKED ISSUE #42877: AIP-84 Migrate the public endpoint Get DAG Stats to FastAPI
### Description

Currently the Get DAG Stats public endpoint is at `api_connexion/endpoints/dag_stats_endpoint.py` under `get_dag_stats`. We need to migrate it to the `api_fastapi/views/public/dag_stats.py` under a `get_dag_stats` or similar.

### Use case/motivation

AIP-84 cWiki Link: https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-84+UI+REST+API

### Related issues

Related to: https://github.com/apache/airflow/issues/42370

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by omkar-foss**
closes: #42877
related: #42370

This migrates the Get DAG Stats API from `api_connexion` to `api_fastapi`.

### CODE REVIEW

**Review by jscheffl — APPROVED** [review]

File: airflow/api_fastapi/core_api/routes/public/dag_stats.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+dag_stats_router = AirflowRouter(tags=["DagStats"], prefix="/dagStats")
+
+
+@dag_stats_router.get(
+    "/",
+    responses=create_openapi_http_exception_doc([400, 401, 403, 404, 422]),
+)
+async def get_dag_stats(
+    session: Annotated[Session, Depends(get_session)],
+    dag_ids: str | None = None,
```
Comment (pierrejeambrun) [inline airflow/api_fastapi/core_api/routes/public/dag_stats.py]: FastAPI handles natively lists. You don't need to do that by hands. `dag_ids: list[str] | None = Query(None)`

Comment (pierrejeambrun) [reply airflow/api_fastapi/core_api/routes/public/dag_stats.py]: You can make a common parameter `paramters` because this filtering based on `dag_ids` list is pretty common in the API.

Comment (pierrejeambrun) [inline airflow/api_fastapi/core_api/routes/public/dag_stats.py]: Why did you remove the `limit` and `offset` params that we had originally in the endpoint ?

File: airflow/api_fastapi/core_api/routes/public/dag_stats.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            dag_id=dag_id,
+            stats=[
+                DagStatsStateResponse(
+                    state=state,
+                    count=dag_state_data.get((dag_id, state), 0),
+                )
+                for state in DagRunState
+            ],
+        )
+        for dag_id in sorted(result_dag_ids)
```
Comment (pierrejeambrun) [inline airflow/api_fastapi/core_api/routes/public/dag_stats.py]: The base implementation of the legacy code looked fine. You can follow it to also construct a dict object, and simply instanciate a `DagStatsCollectionResponse` directly.

Pydantic knows how to go from a dict to pydantic model. The issue we have usually is that we have an ORM instance, therefore we need to play around `from_attributes=True` etc...

**Review by pierrejeambrun — CHANGES_REQUESTED** [review]

File: airflow/api_fastapi/core_api/routes/public/dag_stats.py (hunk 1 above)
Comment (omkar-foss) [reply airflow/api_fastapi/core_api/routes/public/dag_stats.py]: The `limit` and `offset` params in the old API are based on filtering the user requested dag ids intersecting with permitted dag ids. But right now without permissions our new API is directly controlled by `dag_ids` query param, so if user needs to paginate they can do so directly by passing the required `dag_ids` alone. So `limit` and `offset` params are redundant in a way until we add permissions. Let me know if this makes sense, thanks!

Comment (omkar-foss) [reply airflow/api_fastapi/core_api/routes/public/dag_stats.py]: We can do that, but that'll change the request format a bit as follows:

- Current (old API) request: `/dagStats?dag_ids=dag1,dag2`
- In new API, if we use `list[str] | None = Query(None)`, then request will be: `/dagStats?dag_ids=dag1&dag_ids=2` instead of `/dagStats?dag_ids=dag1,dag2`.

Let me know if this change in request format is fine, I'll update it to `dag_id` accordingly and create a common parameter. Thank you! 😃

File: airflow/api_fastapi/core_api/routes/public/dag_stats.py (hunk 2 above)
Comment (omkar-foss) [reply airflow/api_fastapi/core_api/routes/public/dag_stats.py]: Done, updated. Thank you! You're a Pydantic pro 🙌🏽

File: tests/api_fastapi/core_api/routes/public/test_dag_stats.py [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+            dag_1,
+            dag_1_run_1,
+            dag_1_run_2,
+            dag_2,
+            dag_2_run_1,
+            dag_3,
+            dag_3_run_1,
+        )
+        for entity in entities:
+            session.add(entity)
```
Comment (pierrejeambrun) [inline tests/api_fastapi/core_api/routes/public/test_dag_stats.py]:
```suggestion
    session.add_all(entities)
```

File: airflow/api_fastapi/core_api/routes/public/dag_stats.py (hunk 1 above)
Comment (pierrejeambrun) [reply airflow/api_fastapi/core_api/routes/public/dag_stats.py]: Yes, FastAPI handles list natively in the exploded way. This will be one of the API breaking change I believe and I don't think we want to manually handle list to make that backward compatible.

HAving `?param=item1&param=item2` is the way to go with FastAPI.

File: tests/api_fastapi/core_api/routes/public/test_dag_stats.py [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+        assert response.status_code == 200
+        res_json = response.json()
+        assert len(res_json["dags"]) == 2
+        assert sorted(
+            res_json["dags"],
+            key=lambda d: d["dag_id"],
+        ) == sorted(
+            exp_payload["dags"],
+            key=lambda d: d["dag_id"],
+        )
```
Comment (pierrejeambrun) [inline tests/api_fastapi/core_api/routes/public/test_dag_stats.py]: Why are we manually sorting the dag_id ? This should be enforced by the endpoint. A stable order should be returned, that we can test.

File: airflow/api_fastapi/core_api/routes/public/dag_stats.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    query: Query = select(
+        DagRun.dag_id,
+        DagRun.state,
+        func.count(DagRun.state),
+    )
+    if dag_ids:
+        query_dag_ids = sorted(set(dag_ids.split(",")))
+        query = query.where(DagRun.dag_id.in_(query_dag_ids))
+    query = query.group_by(DagRun.dag_id, DagRun.state)
+    query_result = session.execute(query)
```
Comment (pierrejeambrun) [inline airflow/api_fastapi/core_api/routes/public/dag_stats.py]: Here you can add a default `order_by(Dag.dag_id)` in the query

File: airflow/api_fastapi/core_api/routes/public/dag_stats.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+    result_dag_ids = set()
```
Comment (pierrejeambrun) [inline airflow/api_fastapi/core_api/routes/public/dag_stats.py]: instead of using a set here, you can use a `list` and only `append` to it if it does not contain the `id`.

This way you have `unique` `dag_ids` inside it, but the ordering is still maintained, and you response will have the appropriate order because we iterate the result set that will itself be in the correct order.

File: airflow/api_fastapi/core_api/routes/public/dag_stats.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    dag_ids: str | None = None,
+) -> DagStatsCollectionResponse:
+    """Get Dag statistics."""
+    query: Query = select(
+        DagRun.dag_id,
+        DagRun.state,
+        func.count(DagRun.state),
+    )
+    if dag_ids:
+        query_dag_ids = sorted(set(dag_ids.split(",")))
```
Comment (pierrejeambrun) [inline airflow/api_fastapi/core_api/routes/public/dag_stats.py]: Not sure this `sorted` is of any use. The actual order is defined bellow by the `sorted(result_dag_id)` that we can get rid of if we construct the `result_dag_ids` in a sorted way.

**Review by pierrejeambrun — COMMENTED** [review]
Looking good. A few suggestions and we should be good to merge.

File: airflow/api_fastapi/core_api/routes/public/dag_stats.py (hunk 2 above)
Comment (omkar-foss) [reply airflow/api_fastapi/core_api/routes/public/dag_stats.py]: @pierrejeambrun just a heads-up, that although FastAPI infers the types correctly, MyPy throws an error during Breeze checks:
```bash
airflow/api_fastapi/core_api/routes/public/dag_stats.py:80: error: Argument
"dags" to "DagStatsCollectionResponse" has incompatible type
"list[dict[str, Any]]"; expected "list[DagStatsResponse]"  [arg-type]
        return DagStatsCollectionResponse(dags=dags, total_entries=len(dag...
                                               ^~~~
Found 1 error in 1 file (checked 4 source files)
Error 1 returned
```

File: tests/api_fastapi/core_api/routes/public/test_dag_stats.py (hunk 4 above)
Comment (omkar-foss) [reply tests/api_fastapi/core_api/routes/public/test_dag_stats.py]: Thanks, updated PR. Not sure why I didn't notice and remove this extra sorting earlier! 🥲

File: airflow/api_fastapi/core_api/routes/public/dag_stats.py (hunk 5 above)
Comment (omkar-foss) [reply airflow/api_fastapi/core_api/routes/public/dag_stats.py]: Done, added default sort by dag_id.

File: airflow/api_fastapi/core_api/routes/public/dag_stats.py (hunk 6 above)
Comment (omkar-foss) [reply airflow/api_fastapi/core_api/routes/public/dag_stats.py]: Done.

File: airflow/api_fastapi/core_api/routes/public/dag_stats.py (hunk 7 above)
Comment (omkar-foss) [reply airflow/api_fastapi/core_api/routes/public/dag_stats.py]: Done, removed the extra sorting and updated PR.

File: airflow/api_fastapi/core_api/routes/public/dag_stats.py (hunk 1 above)
Comment (omkar-foss) [reply airflow/api_fastapi/core_api/routes/public/dag_stats.py]: @pierrejeambrun let me know if this is fine, I can keep the `limit` and `offset`, but just that it won't serve any purpose until we add permissions.

File: tests/api_fastapi/core_api/routes/public/test_dag_stats.py (hunk 3 above)
Comment (omkar-foss) [reply tests/api_fastapi/core_api/routes/public/test_dag_stats.py]: Done.

File: airflow/api_fastapi/core_api/routes/public/dag_stats.py (hunk 1 above)
Comment (omkar-foss) [reply airflow/api_fastapi/core_api/routes/public/dag_stats.py]: Done.

File: airflow/api_fastapi/core_api/routes/public/dag_stats.py (hunk 2 above)
Comment (omkar-foss) [reply airflow/api_fastapi/core_api/routes/public/dag_stats.py]: MyPy checks failing in GitHub actions as well here: https://github.com/apache/airflow/actions/runs/11557750391/job/32169262772?pr=43255#step:6:118

File: airflow/api_fastapi/core_api/routes/public/dag_stats.py (hunk 1 above)
Comment (omkar-foss) [reply airflow/api_fastapi/core_api/routes/public/dag_stats.py]: >You can make a common parameter paramters because this filtering based on dag_ids list is pretty common in the API.

This done too. Also have moved the query to the db commons module. Updated this PR.

File: airflow/api_fastapi/core_api/routes/public/dag_stats.py (hunk 2 above)
Comment (omkar-foss) [reply airflow/api_fastapi/core_api/routes/public/dag_stats.py]: Reverting to use explicit model constructors so the Breeze checks can pass.

**Comment by omkar-foss** [comment]
@pierrejeambrun All comments resolved, please have a look when possible. Thank you!

File: airflow/api_fastapi/common/db/dags.py [hunk 8]
```diff
@@ -45,3 +45,13 @@
...
+
+dagruns_select_with_state_count = (
+    select(
+        DagRun.dag_id,
+        DagRun.state,
+        func.count(DagRun.state),
+    )
+    .group_by(DagRun.dag_id, DagRun.state)
+    .order_by(DagRun.dag_id)
+)
```
Comment (pierrejeambrun) [inline airflow/api_fastapi/common/db/dags.py]: I assume this should go into a `dag_runs.py` module ? This `dags.py` is supposed to be for common resquests regarding dags.

File: airflow/api_fastapi/core_api/routes/public/dag_stats.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    DagStatsStateResponse,
+)
+from airflow.utils.state import DagRunState
+
+dag_stats_router = AirflowRouter(tags=["DagStats"], prefix="/dagStats")
+
+
+@dag_stats_router.get(
+    "/",
+    responses=create_openapi_http_exception_doc([400, 401, 403, 404, 422]),
```
Comment (pierrejeambrun) [inline airflow/api_fastapi/core_api/routes/public/dag_stats.py]: 422 is automatically handled by fastapi and shouldn't be specified here to construct de swagger doc of error responses.

File: tests/api_fastapi/core_api/routes/public/test_dag_stats.py [hunk 10]
```diff
@@ -0,0 +1,409 @@
...
+                                    "count": 1,
+                                },
+                            ],
+                        }
+                    ],
+                    "total_entries": 1,
+                },
+            ),
+            (
+                f"{API_PREFIX}?dag_ids={DAG3_ID}",
```
Comment (pierrejeambrun) [inline tests/api_fastapi/core_api/routes/public/test_dag_stats.py]: This is hard to read. Don't do the query param encoding manually. Give a dictionnary to `param` keyword of the `test_client` call. (spaces or special characters needs encoding)

**Review by pierrejeambrun — COMMENTED** [review]
We are missing the `@mark_fastapi_migration_done` decorator

File: tests/api_fastapi/core_api/routes/public/test_dag_stats.py (hunk 10 above)
Comment (omkar-foss) [reply tests/api_fastapi/core_api/routes/public/test_dag_stats.py]: Done.

File: airflow/api_fastapi/core_api/routes/public/dag_stats.py (hunk 9 above)
Comment (omkar-foss) [reply airflow/api_fastapi/core_api/routes/public/dag_stats.py]: Done.

File: airflow/api_fastapi/common/db/dags.py (hunk 8 above)
Comment (omkar-foss) [reply airflow/api_fastapi/common/db/dags.py]: Oh okay, I was wondering if we keep all dag-related resources (dag, dag code, dag stats etc.) under same file since we've few queries. But yes, one file per entity works too. Updated PR, moved query to new file `dag_runs.py` 👍🏽

**Comment by omkar-foss** [comment]
> We are missing the `@mark_fastapi_migration_done` decorator

Done, added it. And have added `legacy api` label to PR.

**Comment by omkar-foss** [comment]
@pierrejeambrun all conversations and conflicts resolved, PR ready to review ✅

**Review by pierrejeambrun — APPROVED** [review]
Looking good, thanks

**Comment by pierrejeambrun** [comment]
Need rebasing but can be merged after that.

**Comment by omkar-foss** [comment]
> Need rebasing but can be merged after that.

@pierrejeambrun done, PR rebased and synced with `main` ✅

**Comment by pierrejeambrun** [comment]
Thanks @omkar-foss

**Review by pierrejeambrun — APPROVED** [review]

File: airflow/api_fastapi/core_api/routes/public/dag_stats.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    DagStatsStateResponse,
+)
+from airflow.utils.state import DagRunState
+
+dag_stats_router = AirflowRouter(tags=["DagStats"], prefix="/dagStats")
+
+
+@dag_stats_router.get(
+    "/",
+    responses=create_openapi_http_exception_doc([400, 401, 403, 404]),
```
Comment (pierrejeambrun) [inline airflow/api_fastapi/core_api/routes/public/dag_stats.py:44]: Just a reminder, here we now use constants from `fastapi.status`. I'll merge and fix as there are other endpoints that need adjustement @omkar-foss
